File: gradle-scala-kotlin/lib/src/main/scala/com/sapient/learning/ObjectOriented.scala
package com.sapient.learning

abstract class Vehicle {
  def accelerate(): Unit

  def decelerate(): Unit
}

trait Electric {
  this: Insured =>
  def recharge(): Unit
}

trait Insured {
  def renew(): Unit
}

class Taxi(val make: String, val model: String) extends Vehicle with Electric with Insured {

  def this(make: String) = {
    this(make, "Unknown")
  }

  private val _description = s"($make, $model)"

  {
    import Taxi._
    log(_description)
  }

  def description: String = _description

  def accelerate(): Unit = println("Accelerating")

  def decelerate(): Unit = println("decelerating")

  def recharge(): Unit = println("Recharging")

  def renew(): Unit = println("Renewing")

}

object Taxi {
  private def log(message: String): Unit = Logger.info(message)
}

object Logger {
  def info(message: String): Unit = println(s"INFO: $message")
}

object ObjectOriented {

  def show(implicit message: String): Unit = {
    println(message)
  }

  def main(args: Array[String]): Unit = {
    val myTaxi = new Taxi("Ford", "1986")
    Logger.info(myTaxi.description)
    myTaxi.accelerate()
    myTaxi.decelerate()
    myTaxi.recharge()
    myTaxi.renew()

    implicit val message:String = "Hello World!"
    show
    show("Hi World!")

  }
}

File: gradle-scala-kotlin/lib/src/main/kotlin/com/sapient/learning/ObjectOriented.kt
package com.sapient.learning

abstract class Vehicle {
    abstract fun accelerate()
    abstract fun decelerate()
}

interface Electric {
    fun recharge()
}

fun interface Insured {
    fun renew()
}

class BaseInsured : Insured {
    override fun renew() {
        Logger.info("Renewing")
    }
}

class Bus(insured: Insured) : Vehicle(), Insured by insured {
    override fun accelerate() {
        Logger.info("Accelerating")
    }

    override fun decelerate() {
        Logger.info("Decelerating")
    }
}

val insured = Insured({ println("Renewing") })

object Logger {
    fun info(message: String) {
        println("INFO: $message")
    }
}

class Bike : Vehicle(), Electric, Insured {
    override fun accelerate() {
        log("Accelerating")
    }

    override fun decelerate() {
        log("Decelerating")
    }

    override fun recharge() {
        log("Recharge")
    }

    override fun renew() {
        log("Renew")
    }

    companion object {
        private fun log(message: String) {
            Logger.info(message)
        }
    }
}

fun Bike.start() {
    Logger.info("Starting")
}

class Car(val make: String, val model: String) : Vehicle() {
    private val description = "($make, $model)"

    constructor(make: String) : this(make, "Unknown") {
        print(description)
    }

    override fun accelerate() {}
    override fun decelerate() {}
}


fun main() {

    val car: Car = Car("Ford")

    val bike: Bike = Bike()
    bike.start()

    val bus: Bus = Bus(BaseInsured())

}